### PR TITLE
Defork collectd

### PR DESCRIPTION
**Pull request by nmdayton**
Brings Stackdriver's collectd fork in line with the main collectd repo.

### CODE REVIEW

File: src/processes.c [hunk 1]
```diff
@@ -25,2153 +25,3026 @@
...
-int getargs (struct procentry64 *processBuffer, int bufferLen, char *argsBuffer, int argsLen);
+int getargs(void *processBuffer, int bufferLen, char *argsBuffer, int argsLen);
 #endif /* HAVE_PROCINFO_H */
 
+#if HAVE_LIBTASKSTATS
+static ts_t *taskstats_handle;
+#endif
+
+static derive_t ps_delta(derive_t value) {
+	return (value == -1) ? 0 : value;
```
Comment (jkohen) [inline src/processes.c]: `-1` is used as a magic value throughout the code. Please define a constant (VALUE_UNSET?) and use it whenever these variables are initialized, set, or compared to -1 like here.

File: src/processes.c [hunk 2]
```diff
@@ -25,2153 +25,3026 @@
...
+    value_list_t vl = VALUE_LIST_INIT;
+    vl.values = values;
+
+    sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+    sstrncpy (vl.plugin, "processes", sizeof (vl.plugin));
+    sstrncpy (vl.plugin_instance, instance_name, sizeof (vl.plugin_instance));
+
+    if (doing_detail)
+    {
+        // sstrncpy(vl.type_instance, "detail", sizeof (vl.type_instance));
```
Comment (jkohen) [inline src/processes.c]: Please comment why this is commented out, or remove the line.

File: src/processes.c [hunk 3]
```diff
@@ -25,2153 +25,3026 @@
...
+static int ps_delay(procstat_entry_t *ps) {
+  if (taskstats_handle == NULL) {
+    return ENOTCONN;
+  }
+
+  int status = ts_delay_by_tgid(taskstats_handle, (uint32_t)ps->id, &ps->gauges.delay);
+  if (status == EPERM) {
+    static c_complain_t c;
+#if defined(HAVE_SYS_CAPABILITY_H) && defined(CAP_NET_ADMIN)
+    if (check_capabil
```
Comment (jkohen) [inline src/processes.c]: There seems to be inconsistent identation in this file (the two lines above are one instance), perhaps due to mixed usage of spaces and tabs. Can you please unify it following the upstream convention?

**Review by jkohen — APPROVED** [review]
Excellent work!

As discussed offline, I focused the review on processes.c, especially the changes against collectd HEAD (https://github.com/collectd/collectd/compare/master...Stackdriver:nina-marie-collectd)

You suggested to look at src/Makefile.am, but I can't make head or tails of it. IT's a large change, but if it works, go for it.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]
Should we not be reviewing the diff between our current master (i.e., `stackdriver-agent-5.5.2`) and the proposed HEAD, so we know what we're picking up?
Eventually, this should be our forked changes rebased on top of the upstream `master`, and we could review that as well, but first we need to know what's changing in our copy, right?
I'd propose sitting down with @nina-marie when I'm back and rebasing these changes together.

**Comment by bsiegert** [comment]
How feasible would it be to get rid of the fork altogether and have the `write_gcm` plugin (which, I gather, is what makes the stackdriver integration work at all) be a standalone repo, requiring an upstream collectd installed?

(As a packager, this would be much nicer, given that we already have a working collectd port in pkgsrc.)

**Comment by igorpeshansky** [comment]
@bsiegert Unfortunately, this isn't how collectd plugins work. They are required to be built along with the collectd binary (except for some rare cases like plugins written in another language).
We've toyed with the idea of maintaining a separate repo for the `write_gcm` plugin and pulling in upstream collectd via a git submodule, but the work in this PR is a prerequisite for that happening in any case. Another possibility is to converge upstream's `write_stackdriver` and our `write_gcm` to the point where we can simply use the upstream codebase. At the moment, however, none of those are actual planned work.

**Comment by igorpeshansky** [comment]
Closing in favor of #150.

**Pull request closed without merging**